File: game-logic-core/scala3MainProject/src/main/scala/fwc/game/actions/Action.scala
package fwc.game.actions

import fwc.JsonSerializable
import fwc.game.GameState
import fwc.game.actions.action.*
import fwc.game.actions.planning.*
import fwc.game.actions.roundEvents.*


trait Action(gameState: GameState) extends JsonSerializable {
  def doAction(): GameState
  def undoAction(): GameState = gameState
}

object Action extends JsonParsableAction {
  val actionTypeJsonKey = "actionType"

  //todo refactor the rest of cases
  enum PlanningActions(val string: String) {
    case AddOrder extends PlanningActions("addOrder")
    case RemoveOrder extends PlanningActions("removeOrder")
    case OpenOrders extends PlanningActions("openOrders")
  }

  def fromJson(gameState: GameState, json: ujson.Value): Action = {
    json(actionTypeJsonKey).str match
      case PlanningActions.AddOrder.string => ActionAddOrder.fromJson(gameState, json)
      case PlanningActions.RemoveOrder.string => ActionRemoveOrder.fromJson(gameState, json)
      case PlanningActions.OpenOrders.string => ActionOpenOrders.fromJson(gameState, json)
      case "ravenChooseChangeOrderOrLookAtWildlingCard" => ActionRavenChooseChangeOrderOrLookAtWildlingCard.fromJson(gameState, json)
      case "ravenChangeOrder" => ActionRavenChangeOrder.fromJson(gameState, json)
      case "ravenChoosePutWildlingsCardOnTopOrBottom" => ActionRavenChoosePutWildlingsCardOnTopOrBottom.fromJson(gameState, json)
      case "ravenGetWildlingsCard" => ActionRavenGetWildlingsCard.fromJson(gameState, json)

      case "resolveRaidOrder" => ActionResolveRaidOrder.fromJson(gameState, json)
      case "resolveMarchOrder" => ActionResolveMarchOrder.fromJson(gameState, json)
      case "leavePowerTokenAtTile" => ActionLeavePowerTokenAtTile.fromJson(gameState, json)
      case "resolveSupportOrder" => ActionResolveSupportOrder.fromJson(gameState, json)
      case "chooseHouseCard" => ActionChooseHouseCard.fromJson(gameState, json)
      case "resolveCardPufferfish0" => ActionResolveCardPufferfish0.fromJson(gameState, json)
      case "resolveCardKraken6" => ActionResolveCardKraken6.fromJson(gameState, json)
      case "resolveCardRose2" => ActionResolveCardRose2.fromJson(gameState, json)
      case "resolveCardRose4" => ActionResolveCardRose4.fromJson(gameState, json)
      case "resolveCardLion5" => ActionResolveCardLion5.fromJson(gameState, json)
      case "getTidesOfBattleCards" => ActionGetTidesOfBattleCards.fromJson(gameState, json)
      case "setTidesOfBattleCards" => ActionSetTidesOfBattleCards.fromJson(gameState, json)
      case "refreshTidesOfBattleDeck" => ActionRefreshTidesOfBattleDeck.fromJson(gameState, json)
      case "useValyrianSteelBlade" => ActionUseValyrianSteelBlade.fromJson(gameState, json)
      case "calculateCombatOutcome" => ActionCalculateCombatOutcome.fromJson(gameState, json)
      case "resolveCardWolf0" => ActionResolveCardWolf0.fromJson(gameState, json)
      case "killUnitsAfterBattle" => ActionKillUnitsAfterBattle.fromJson(gameState, json)
      case "retreatUnitsAfterBattle" => ActionRetreatUnitsAfterBattle.fromJson(gameState, json)
      case "disbandUnitsAfterCombat" => ActionDisbandUnitsAfterCombat.fromJson(gameState, json)
      case "cleanUpAfterCombat" => ActionCleanUpAfterCombat.fromJson(gameState, json)
      case "resolveCardMoose2" => ActionResolveCardMoose2.fromJson(gameState, json)
      case "resolveCardLion1" => ActionResolveCardLion1.fromJson(gameState, json)
      case "chooseHouseCardAfterLion5" => ActionChooseHouseCardAfterLion5.fromJson(gameState, json)
      case "resolveCardMoose3" => ActionResolveCardMoose3.fromJson(gameState, json)
      case "resolveSpecialConsolidatePower" => ActionResolveSpecialConsolidatePower.fromJson(gameState, json)
      case "resolveConsolidatePowerOrder" => ActionResolveConsolidatePowerOrder.fromJson(gameState, json)
      case "cleanUpAfterRound" => ActionCleanUpAfterRound.fromJson(gameState, json)

      case "calculateGameWinner" => ActionCalculateGameWinner.fromJson(gameState, json)

      case "getEventCards" => ActionGetEventCards.fromJson(gameState, json)
      case "setEventCards" => ActionSetEventCards.fromJson(gameState, json)
      case "getWildlingsCard" => ActionGetWildlingsCard.fromJson(gameState, json)
      case "setWildlingsCard" => ActionSetWildlingsCard.fromJson(gameState, json)
      case "recalculateSupplies" => ActionRecalculateSupplies.fromJson(gameState, json)
      case "disbandUnitDueToSupplies" => ActionDisbandUnitDueToSupplies.fromJson(gameState, json)
      case "muster" => ActionMuster.fromJson(gameState, json)
      case "finishMustering" => ActionFinishMustering.fromJson(gameState, json)
      case "stopMustering" => ActionStopMustering.fromJson(gameState, json)
      case "throneChooseSupplyOrMuster" => ActionThroneChooseSupplyOrMuster.fromJson(gameState, json)
      case "collectTaxes" => ActionCollectTaxes.fromJson(gameState, json)
      case "ravenChooseTrackBidsOrCollectTaxes" => ActionRavenChooseTrackBidsOrCollectTaxes.fromJson(gameState, json)
      case "trackBids" => ActionTrackBids.fromJson(gameState, json)
      case "openTrackBids" => ActionOpenTrackBids.fromJson(gameState, json)
      case "resolveTiesAfterBiddingOnTracks" => ActionResolveTiesAfterBiddingOnTracks.fromJson(gameState, json)
      case "disableOrder" => ActionDisableOrder.fromJson(gameState, json)
      case "steelBladeChooseDisableMarchOrDefend" => ActionSteelBladeChooseDisableMarchOrDefend.fromJson(gameState, json)
      case "wildlingsBids" => ActionWildlingsBids.fromJson(gameState, json)
      case "resolveTiesAfterBiddingOnWildlings" => ActionResolveTiesAfterBiddingOnWildlings.fromJson(gameState, json)
      case "wildlingsKillUnit" => ActionWildlingsKillUnit.fromJson(gameState, json)
      case "wildlingsChooseKill2UnitsOr2PositionsOnTrack" => ActionWildlingsChooseKill2UnitsOr2PositionsOnTrack.fromJson(gameState, json)
      case "wildlingsDiscardHouseCard" => ActionWildlingsDiscardHouseCard.fromJson(gameState, json)
      case "wildlingsReturnHouseCard" => ActionWildlingsReturnHouseCard.fromJson(gameState, json)
      case "wildlingsCard" => ActionWildlingsCard.fromJson(gameState, json)
      case "wildlingsMusterAtCastle" => ActionWildlingsMusterAtCastle.fromJson(gameState, json)
      case "wildlingsDowngradeKnights" => ActionWildlingsDowngradeKnights.fromJson(gameState, json)
      case "wildlingsUpgradeKnights" => ActionWildlingsUpgradeKnights.fromJson(gameState, json)
      case "wildlingsChooseTrackToBeFirstAt" => ActionWildlingsChooseTrackToBeFirstAt.fromJson(gameState, json)
      case "wildlingsChooseTrackToBeLastAt" => ActionWildlingsChooseTrackToBeLastAt.fromJson(gameState, json)
      case "autoKillUnitsAfterBattle" => ActionAutoKillUnitsAfterBattle.fromJson(gameState, json)
      case "autoRetreatAfterBattle" => ActionAutoRetreatAfterBattle.fromJson(gameState, json)

  }
}

File: bots-di/DTO/actions/action.py
from typing import TypedDict, Optional, NotRequired

from server_module.game_state.military_unit import HouseType, MilitaryUnit
from server_module.game_state.track_type import TrackType


class ActionResolveCardLion1(TypedDict):
    actionType: str  # = "resolveCardLion1"
    houseType: HouseType
    tileNumber: int


class ActionDisbandUnitsAfterCombat(TypedDict):
    actionType: str  # = "disbandUnitsAfterCombat"
    houseType: HouseType
    unit: MilitaryUnit


class ActionResolveCardMoose3(TypedDict):
    actionType: str  # = "resolveCardMoose3"
    houseType: HouseType
    cardCode: int


class ActionRetreatUnitsAfterBattle(TypedDict):
    actionType: str  # = "retreatUnitsAfterBattle"
    houseType: HouseType
    targetTileNumber: int


class ActionResolveCardWolf0(TypedDict):
    actionType: str  # = "resolveCardWolf0"
    houseType: HouseType
    targetTileNumber: int


class ActionResolveCardMoose2(TypedDict):
    actionType: str  # = "resolveCardMoose2"
    houseType: HouseType
    tileNumber: int


class ActionUseValyrianSteelBlade(TypedDict):
    actionType: str  # = "useValyrianSteelBlade"
    houseType: HouseType
    choice: str


class ActionRefreshTidesOfBattleDeck(TypedDict):
    actionType: str  # = "refreshTidesOfBattleDeck"
    newCards: list[int]

class AfterCombatStateJson(TypedDict):
    armies: dict
    discardedHouseCards: dict
    placedOrders: dict
    powerTokens: dict

class ActionCleanUpAfterCombat(TypedDict):
    actionType: str  # = "cleanUpAfterCombat"
    state: AfterCombatStateJson
    doCardResolve: bool


class ActionResolveCardKraken6(TypedDict):
    actionType: str  # = "resolveCardKraken6"
    houseType: HouseType
    newCardCode: int


class ActionCalculateGameWinner(TypedDict):
    actionType: str  # = "calculateGameWinner"


class ActionResolveCardRose4(TypedDict):
    actionType: str  # = "resolveCardRose4"
    houseType: HouseType
    tileNumber: int


class ActionChooseHouseCard(TypedDict):
    actionType: str  # = "chooseHouseCard"
    houseType: HouseType
    cardCode: int


class ActionResolveCardLion5(TypedDict):
    actionType: str  # = "resolveCardLion5"
    houseType: HouseType
    doCancelCard: bool


class ActionResolveSupportOrder(TypedDict):
    actionType: str  # = "resolveSupportOrder"
    fromHouseType: HouseType
    toHouseType: HouseType
    tileNumbers: list[int]


class ActionResolveCardPufferfish0(TypedDict):
    actionType: str  # = "resolveCardPufferfish0"
    houseType: HouseType
    trackType: TrackType


class ActionCalculateCombatOutcome(TypedDict):
    actionType: str  # = "calculateCombatOutcome"


class ActionResolveConsolidatePowerOrder(TypedDict):
    actionType: str  # = "resolveConsolidatePowerOrder"
    powerTokens: dict[HouseType | str, int]


class ActionGetTidesOfBattleCards(TypedDict):
    actionType: str  # = "getTidesOfBattleCards"
    isRandom: bool


class ActionCleanUpAfterRound(TypedDict):
    actionType: str  # = "cleanUpAfterRound"
    isRandom: bool
    round: int


class ActionResolveCardRose2(TypedDict):
    actionType: str  # = "resolveCardRose2"
    houseType: HouseType


class ActionResolveSpecialConsolidatePower(TypedDict):
    actionType: str  # = "resolveSpecialConsolidatePower"
    houseType: HouseType
    fromTile: int
    toTile: NotRequired[int]
    unitToMuster: NotRequired[MilitaryUnit]
    isUpgrade: bool


class ActionSetTidesOfBattleCards(TypedDict):
    actionType: str  # = "setTidesOfBattleCards"
    attackerCardCode: int
    defenderCardCode: int


class ActionKillUnitsAfterBattle(TypedDict):
    actionType: str  # = "killUnitsAfterBattle"
    houseType: HouseType
    units: list[MilitaryUnit]


class ActionLeavePowerTokenAtTile(TypedDict):
    actionType: str  # = "leavePowerTokenAtTile"
    houseType: HouseType
    doLeave: bool
    tileNumber: int


class ActionResolveMarchOrder(TypedDict):
    actionType: str  # = "resolveMarchOrder"
    houseType: HouseType
    sourceTileNumber: int
    targets: dict[int, list[MilitaryUnit]]


class ActionResolveRaidOrder(TypedDict):
    actionType: str  # = "resolveRaidOrder"
    houseType: HouseType
    sourceTileNumber: int
    targetTileNumber: int


class ActionChooseHouseCardAfterLion5(TypedDict):
    actionType: str  # = "chooseHouseCardAfterLion5"
    houseType: HouseType
    cardCode: int
